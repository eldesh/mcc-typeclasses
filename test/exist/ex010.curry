-- Here is a rather contrived example combining universal and exisential
-- quantification with constraints. Note that the Functor constraint on
-- T is not stricly necessary (in fact, any constraint on universally
-- quantified variables in a data type declaration is not necessary).

data Functor f => T f = forall a. (Enum a, Show (f a)) => T (f a)

instance Functor f => Show (T f) where
  showsPrec p (T xs) = showsPrec p xs

tenum :: Functor f => T f -> f Int
tenum (T xs) = fmap fromEnum xs

main =
  do
    let ts = [T [1,2,3], T "hello"]
    putStrLn (show ts ++ " => " ++ show (map tenum ts))
    let t' = T (Just ())
    putStrLn (show t' ++ " => " ++ show (tenum t'))
