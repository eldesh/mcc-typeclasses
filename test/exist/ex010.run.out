[[1,2,3],"hello"] => [[1,2,3],[104,101,108,108,111]]
Just () => Just 0
